Shopping cart

ShastkivRuslan opened the pull request:
Implement Shopping Cart Management with CRUD endpoints and validation

- Add functionality for users to manage their shopping cart with the following endpoints:
  - POST /api/cart: Add a book to the cart
  - GET /api/cart: Retrieve the shopping cart
  - PUT /api/cart/items/{cartItemId}: Update book quantity in cart
  - DELETE /api/cart/items/{cartItemId}: Remove a book from the cart

- Implement domain models: ShoppingCart, CartItem
- Establish relationships between entities with soft delete for ShoppingCart
- Configure cascading and validation for input DTOs
- Add Liquibase for database migrations, avoiding FetchType.EAGER
- Integrate Swagger for API documentation and enable pagination and sorting
- Ensure ShoppingCart creation during user registration and apply role-based access control to endpoints

ShastkivRuslan committed:
Made some improvements to the previous task:
-Solved the N+1 problem when getting all books with categories from the DB;
-Refactor BookDto to return Set<CategoryDto> instead of Set<Category>, replacing @JsonIgnore usage;
-Add check to filter out deleted books when retrieving all books for a specific category;

ShastkivRuslan committed:
Added ShoppingCart and CartItem entities with the required fields and relations;

ShastkivRuslan committed:
Added Liquibase scripts to create required tables;

ShastkivRuslan committed:
Added required dto`s;

ShastkivRuslan committed:
-Enabled creation of a shopping cart upon new user registration

ShastkivRuslan committed:
Most of the functionality has been implemented.

ShastkivRuslan committed:
completed;

Elena-Bruyako commented on src/main/java/ruslan/shastkiv/bookstore/controller/ShoppingCartController.java:
```diff
@@ -0,0 +1,72 @@
+    @PreAuthorize("hasRole('USER')")
+    @ResponseStatus(HttpStatus.CREATED)
+    @Operation(summary = "Add a book to the shopping cart")
+    public CartItemDto addBook(
```
```suggestion
    public ShoppingCartDto addBook(
```

Elena-Bruyako commented on src/main/java/ruslan/shastkiv/bookstore/controller/ShoppingCartController.java:
```diff
@@ -0,0 +1,72 @@
+    @PutMapping("/items/{cartItemId}")
+    @PreAuthorize("hasRole('USER')")
+    @Operation(summary = "Update quantity of a cart item")
+    public CartItemDto updateQuantity(
```
```suggestion
    public ShoppingCartDto updateQuantity(
```

Elena-Bruyako commented on src/main/java/ruslan/shastkiv/bookstore/controller/ShoppingCartController.java:
```diff
@@ -0,0 +1,72 @@
+@Tag(name = "Shopping Cart", description = "Endpoints for managing the shopping cart")
+public class ShoppingCartController {
+    private final ShoppingCartService shoppingCartService;
+    private final CartItemService cartItemService;
```
```suggestion
```

Elena-Bruyako commented on src/main/java/ruslan/shastkiv/bookstore/dto/item/CartItemRequestDto.java:
```diff
@@ -0,0 +1,12 @@
+public record CartItemRequestDto(@NotNull(message = "Book ID cannot be null")
+                                 @Positive(message = "Book ID must be greater than 0")
+                                 Long bookId,
+                                 @Min(value = 1, message = "Quantity must be at least 1")
```
It's better to use `@Positive`

Elena-Bruyako commented on src/main/java/ruslan/shastkiv/bookstore/dto/item/UpdateCartItemRequestDto.java:
```diff
@@ -0,0 +1,7 @@
+
+import jakarta.validation.constraints.Min;
+
+public record UpdateCartItemRequestDto(@Min(value = 1, message = "Quantity must be at least 1")
```
same here

Elena-Bruyako commented on src/main/java/ruslan/shastkiv/bookstore/model/CartItem.java:
```diff
@@ -0,0 +1,33 @@
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    @ManyToOne
```
```suggestion
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
```

Elena-Bruyako commented on src/main/java/ruslan/shastkiv/bookstore/model/CartItem.java:
```diff
@@ -0,0 +1,33 @@
+    @JoinColumn(name = "shopping_cart_id", nullable = false)
+    private ShoppingCart shoppingCart;
+
+    @ManyToOne
```
```suggestion
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
```

Elena-Bruyako commented on src/main/java/ruslan/shastkiv/bookstore/model/ShoppingCart.java:
```diff
@@ -0,0 +1,36 @@
+    private Long id;
+
+    @OneToOne(cascade = CascadeType.REMOVE)
+    @MapsId
```
```suggestion
    @MapsId
    @JoinColumn(name = "id", nullable = false)
```

Elena-Bruyako commented on src/main/java/ruslan/shastkiv/bookstore/model/ShoppingCart.java:
```diff
@@ -0,0 +1,36 @@
+    @Id
+    private Long id;
+
+    @OneToOne(cascade = CascadeType.REMOVE)
```
```suggestion
    @OneToOne(fetch = FetchType.LAZY)
```

Elena-Bruyako commented on src/main/java/ruslan/shastkiv/bookstore/model/ShoppingCart.java:
```diff
@@ -0,0 +1,36 @@
+    @MapsId
+    private User user;
+
+    @OneToMany(mappedBy = "shoppingCart", cascade = CascadeType.ALL)
```
```suggestion
    @OneToMany(mappedBy = "shoppingCart", cascade = CascadeType.ALL, orphanRemoval = true)
```

Elena-Bruyako commented on src/main/java/ruslan/shastkiv/bookstore/model/ShoppingCart.java:
```diff
@@ -0,0 +1,36 @@
+    private User user;
+
+    @OneToMany(mappedBy = "shoppingCart", cascade = CascadeType.ALL)
+    private Set<CartItem> cartItems;
```
```suggestion
    private Set<CartItem> cartItems = new HashSet<>();
```

Elena-Bruyako commented on src/main/java/ruslan/shastkiv/bookstore/service/cart/ShoppingCartService.java:
```diff
@@ -0,0 +1,18 @@
+
+    ShoppingCartDto getShoppingCart(Long userId);
+
+    CartItemDto addBookToCart(Long userId, @Valid CartItemRequestDto requestDto);
```
```suggestion
    CartItemDto addBookToCart(Long userId, CartItemRequestDto requestDto);
```

Elena-Bruyako commented on src/main/java/ruslan/shastkiv/bookstore/service/cart/ShoppingCartServiceImpl.java:
```diff
@@ -0,0 +1,47 @@
+        return cartItemMapper.toDto(cartItemService.addCartItem(shoppingCart, requestDto));
+    }
+
+    @Override
+    public ShoppingCart findShoppingCart(Long id) {
```
```suggestion
    private ShoppingCart findShoppingCart(Long id) {
```

Elena-Bruyako commented on src/main/java/ruslan/shastkiv/bookstore/service/item/CartItemService.java:
```diff
@@ -0,0 +1,18 @@
+import ruslan.shastkiv.bookstore.model.CartItem;
+import ruslan.shastkiv.bookstore.model.ShoppingCart;
+
+public interface CartItemService {
```
all logic should be in ShoppingCartService,  remove CartItemService

Elena-Bruyako commented on src/main/java/ruslan/shastkiv/bookstore/service/user/UserServiceImpl.java:
```diff
@@ -21,6 +23,7 @@ public class UserServiceImpl implements UserService {
     private final RoleRepository roleRepository;
+    private final ShoppingCartService shoppingCartService;
 
     @Override
```
```suggestion
    @Override
    @Transactional
```

Elena-Bruyako reviewed (changes requested):
see comments

ShastkivRuslan committed:
changes after mentor review:
- Changed return type of `addBook` and `updateQuantity` methods in `ShoppingCartController` to `ShoppingCartDto` instead of `CartItemDto`.
- Removed the unused `CartItemService` and moved its logic into `ShoppingCartService`.
- Updated DTOs: replaced `@Min` with `@Positive` for quantity validation in `CartItemRequestDto` and `UpdateCartItemRequestDto`.
- Improved `CartItem` and `ShoppingCart` models:
  - Added `fetch = FetchType.LAZY` and `optional = false` for better lazy loading and null handling.
  - Added `orphanRemoval = true` in `@OneToMany` relationship to ensure proper cascade deletion of cart items.
  - Initialized `cartItems` set to `HashSet` for proper collection handling.
- Adjusted `ShoppingCart` and `CartItem` mappings for improved entity relationships.

ShastkivRuslan commented:
changes after mentor review:
- Changed return type of `addBook` and `updateQuantity` methods in `ShoppingCartController` to `ShoppingCartDto` instead of `CartItemDto`.
- Removed the unused `CartItemService` and moved its logic into `ShoppingCartService`.
- Updated DTOs: replaced `@Min` with `@Positive` for quantity validation in `CartItemRequestDto` and `UpdateCartItemRequestDto`.
- Improved `CartItem` and `ShoppingCart` models:
  -  Added `fetch = FetchType.LAZY` and `optional = false` for better lazy loading and null handling.
  -  Added `orphanRemoval = true` in `@OneToMany` relationship to ensure proper cascade deletion of cart items.
  -  Initialized `cartItems` set to `HashSet` for proper collection handling.
-  Adjusted `ShoppingCart` and `CartItem` mappings for improved entity relationships.

Refactor addCartItem method to simplify logic for adding/updating cart items
- Optimized handling of existing cart items by directly assigning the CartItem.
- Removed unnecessary conditional checks for improved readability.
- Updated item quantity or added new item to shopping cart more efficiently.

ahoienko commented on src/main/java/ruslan/shastkiv/bookstore/mapper/BookMapper.java:
```diff
@@ -1,22 +1,33 @@
+        book.setId(bookId);
+        return book;
+    }
+
```
```suggestion
```

ahoienko commented on src/main/java/ruslan/shastkiv/bookstore/service/cart/ShoppingCartServiceImpl.java:
```diff
@@ -0,0 +1,87 @@
+                                + " ] or Shopping Cart for user ID [ " + userId + " ] not found"));
+    }
+
+    private ShoppingCart addCartItem(ShoppingCart shoppingCart, CartItemRequestDto requestDto) {
+        CartItem cartItem = cartItemRepository
+                .findByBookIdAndShoppingCartId(requestDto.bookId(), shoppingCart.getId())
+                .orElseGet(() -> {
+                    CartItem newCartItem = new CartItem();
+                    newCartItem.setBook(bookService.findBookById(requestDto.bookId()));
+                    newCartItem.setShoppingCart(shoppingCart);
+                    shoppingCart.getCartItems().add(newCartItem);
+                    return newCartItem;
+                });
+        cartItem.setQuantity(cartItem.getQuantity() + requestDto.quantity());
+        return shoppingCartRepository.save(shoppingCart);
```
I think you should not update quantity, because It won`t be possible to decrease amount of requested book in cart

ahoienko commented on src/main/java/ruslan/shastkiv/bookstore/service/cart/ShoppingCartServiceImpl.java:
```diff
@@ -0,0 +1,87 @@
+
+    private ShoppingCart findShoppingCart(Long id) {
+        return shoppingCartRepository.findById(id).orElseThrow(()
+                -> new EntityNotFoundException("Cant find shopping cart by user id: " + id));
+    }
+
+    private CartItem getByCartAndUserId(Long userId, Long cartItemId) {
+        return cartItemRepository.findByIdAndShoppingCartId(cartItemId, userId)
+                .orElseThrow(() -> new EntityNotFoundException(
+                        "Cart item with ID [ " + cartItemId
+                                + " ] or Shopping Cart for user ID [ " + userId + " ] not found"));
```
Why style of error message is different, please be consistent and decide which style to use

ShastkivRuslan replied:
fixed

ahoienko commented on src/main/resources/db/changelog/changes/13-create-shopping_cart-table.yaml:
```diff
@@ -0,0 +1,27 @@
+                  type: bigint
+                  constraints:
+                    primaryKey: true
+                    nullable: false
```
primaryKey: true already means that nullable: false, so no need to add this explicitly

ShastkivRuslan replied:
Reason: liquibase.exception.DatabaseException: Column "USER_ID" must not be nullable

ahoienko commented on src/main/java/ruslan/shastkiv/bookstore/service/cart/ShoppingCartServiceImpl.java:
```diff
@@ -0,0 +1,87 @@
+                -> new EntityNotFoundException("Cant find shopping cart by user id: " + id));
+    }
+
+    private CartItem getByCartAndUserId(Long userId, Long cartItemId) {
```
```suggestion
    private CartItem getByCartItemIdAndUserId(Long userId, Long cartItemId) {
```

ahoienko commented on src/main/java/ruslan/shastkiv/bookstore/controller/ShoppingCartController.java:
```diff
@@ -0,0 +1,69 @@
+            @PathVariable Long cartItemId,
+            Authentication authentication,
+            @RequestBody @Valid UpdateCartItemRequestDto requestDto) {
+        return shoppingCartService.updateItemQuantity(
+                userService.getUserId(authentication), cartItemId, requestDto);
```
In current implementation any user can change quantity of cart item for any other user

ShastkivRuslan replied:
I've tested the implementation, and everything works as expected. Users cannot modify cart items for other users.

ahoienko reviewed (changes requested):
Nice, left some comments

ShastkivRuslan committed:
changes after second mentor review;
changed names of some methods;
removed unnecessary constraints;
refactored the style of error messages;

ShastkivRuslan committed:
Reverted nullable=false on all primary keys to fix application startup issue after deletion.
